File: src/Optimization/StructureAware/HierarchicalApplicator.cs
using Roguelike.Data;
using Roguelike.Core;
using Roguelike.Core.Map;
using System;
using System.Linq;

namespace Roguelike.Optimization
{
    /// <summary>
    /// Applies a hierarchical genome to game data using multiplicative scaling layers
    /// </summary>
    public static class HierarchicalApplicator
    {
        public static void Apply(
            HierarchicalGenome genome, 
            CardPool cardPool, 
            EnemyPool enemyPool, 
            RelicPool relicPool,
            HeroData hero,
            int currentFloor = 1)
        {
            ApplyToHero(genome, hero);
            ApplyToCards(genome, cardPool, currentFloor);
            ApplyToEnemies(genome, enemyPool, currentFloor);
            ApplyToEconomy(genome, cardPool, relicPool);
        }

        private static void ApplyToHero(HierarchicalGenome genome, HeroData hero)
        {
            // Scale hero health
            hero.StartingHealth = (int)Math.Round(hero.StartingHealth * genome.HeroHealthScalar);
            hero.StartingHealth = Math.Max(30, hero.StartingHealth); // Min 30 HP
            
            // Scale starting gold
            hero.StartingGold = (int)Math.Round(hero.StartingGold * genome.HeroStartGoldScalar);
            hero.StartingGold = Math.Max(50, hero.StartingGold); // Min 50 gold
            
            // Adjust mana
            hero.StartingMana += genome.HeroManaOffset;
            hero.StartingMana = Math.Max(2, Math.Min(4, hero.StartingMana)); // 2-4 mana
        }

        private static void ApplyToCards(HierarchicalGenome genome, CardPool cardPool, int floor)
        {
            foreach (var card in cardPool.CardsById.Values)
            {
                for (int i = 0; i < card.Actions.Count; i++)
                {
                    var action = card.Actions[i];
                    
                    switch (action.Type)
                    {
                        case ActionType.DealDamage:
                            action.Value = CalculateCardDamage(genome, card, action.Value, floor);
                            break;
                            
                        case ActionType.GainBlock:
                            action.Value = CalculateCardBlock(genome, card, action.Value, floor);
                            break;
                            
                        case ActionType.ApplyStatusEffect:
                            action.Value = CalculateCardEffectStacks(genome, card, action.Value);
                            break;
                    }
                }
                
                int originalCost = card.ManaCost;
                card.ManaCost = CalculateCardCost(genome, card, originalCost);
            }
        }

        private static int CalculateCardDamage(
            HierarchicalGenome genome, 
            CardData card, 
            int baseValue, 
            int floor)
        {
            float scalar = genome.GlobalDamageMultiplier;
            scalar *= genome.CardTypeScalars[card.Type];
            scalar *= genome.CardStarScalars[card.StarRating];
            scalar *= genome.GetProgressionScalar(floor, HierarchicalGenome.ScalingType.Damage);
            
            if (genome.CardDamageOverrides.TryGetValue(card.Id, out float ovr))
                scalar *= ovr;
            
            return Math.Max(1, (int)Math.Round(baseValue * scalar));
        }

        private static int CalculateCardBlock(
            HierarchicalGenome genome, 
            CardData card, 
            int baseValue, 
            int floor)
        {
            float scalar = genome.GlobalBlockMultiplier;
            scalar *= genome.CardTypeScalars[card.Type];
            scalar *= genome.CardStarScalars[card.StarRating];
            scalar *= genome.GetProgressionScalar(floor, HierarchicalGenome.ScalingType.Block);
            
            return Math.Max(1, (int)Math.Round(baseValue * scalar));
        }

        private static int CalculateCardEffectStacks(
            HierarchicalGenome genome,
            CardData card,
            int baseValue)
        {
            float scalar = genome.CardStarScalars[card.StarRating];
            return Math.Max(1, (int)Math.Round(baseValue * scalar));
        }

        private static int CalculateCardCost(
            HierarchicalGenome genome,
            CardData card,
            int baseCost)
        {
            float scalar = genome.GlobalManaCostMultiplier;
            
            if (genome.CardManaCostOverrides.TryGetValue(card.Id, out float ovr))
                scalar *= ovr;
            
            int newCost = (int)Math.Round(baseCost * scalar);
            return Math.Max(0, newCost);
        }

        private static void ApplyToEnemies(HierarchicalGenome genome, EnemyPool enemyPool, int floor)
        {
            foreach (var enemy in enemyPool.EnemiesById.Values)
            {
                float healthScalar = genome.GlobalHealthMultiplier;
                healthScalar *= genome.EnemyStarScalars[enemy.StarRating];
                healthScalar *= genome.GetProgressionScalar(floor, HierarchicalGenome.ScalingType.Health);
                
                if (genome.EnemyHealthOverrides.TryGetValue(enemy.Id, out float hpOvr))
                    healthScalar *= hpOvr;
                
                enemy.StartingHealth = Math.Max(1, (int)Math.Round(enemy.StartingHealth * healthScalar));
                
                foreach (var weightedAction in enemy.ActionSet)
                {
                    var action = weightedAction.Item;
                    
                    switch (action.Type)
                    {
                        case ActionType.DealDamage:
                            float dmgScalar = genome.GlobalDamageMultiplier;
                            dmgScalar *= genome.EnemyStarScalars[enemy.StarRating];
                            dmgScalar *= genome.GetProgressionScalar(floor, HierarchicalGenome.ScalingType.Damage);
                            action.Value = Math.Max(1, (int)Math.Round(action.Value * dmgScalar));
                            break;
                            
                        case ActionType.GainBlock:
                            float blockScalar = genome.GlobalBlockMultiplier;
                            blockScalar *= genome.EnemyStarScalars[enemy.StarRating];
                            blockScalar *= genome.GetProgressionScalar(floor, HierarchicalGenome.ScalingType.Block);
                            action.Value = Math.Max(1, (int)Math.Round(action.Value * blockScalar));
                            break;
                    }
                }
            }
        }

        private static void ApplyToEconomy(
            HierarchicalGenome genome, 
            CardPool cardPool, 
            RelicPool relicPool)
        {
            for (int star = 1; star <= 5; star++)
            {
                if (cardPool.BaseShopCosts.ContainsKey(star))
                {
                    int baseCost = cardPool.BaseShopCosts[star];
                    cardPool.BaseShopCosts[star] = (int)Math.Round(baseCost * genome.GlobalGoldMultiplier);
                }
                
                if (relicPool.BaseShopCosts.ContainsKey(star))
                {
                    int baseCost = relicPool.BaseShopCosts[star];
                    relicPool.BaseShopCosts[star] = (int)Math.Round(baseCost * genome.GlobalGoldMultiplier);
                }
            }
        }

        /// <summary>
        /// Apply genome specifically for combat encounters (dynamic scaling based on current floor)
        /// </summary>
        public static void ApplyForFloor(
            HierarchicalGenome genome,
            CardPool cardPool,
            EnemyPool enemyPool,
            int floor)
        {
            ApplyToCards(genome, cardPool, floor);
            ApplyToEnemies(genome, enemyPool, floor);
        }
    }
}

File: python_roguelike/optimization/hierarchical_applicator.py
"""
hierarchical_applicator.py  –  Python port of HierarchicalApplicator.cs

Applies a HierarchicalGenome to cloned game data using multiplicative scaling.
"""

import copy
import math
from typing import Dict

from .hierarchical_genome import HierarchicalGenome, ScalingType
from ..data.enums import ActionType, RoomType, CardType
from ..data.pools.card_pool import CardPool
from ..data.pools.relic_pool import RelicPool
from ..data.pools.enemy_pool import EnemyPool
from ..data.combatant.hero_data import HeroData


def apply(
    genome: HierarchicalGenome,
    card_pool: CardPool,
    enemy_pool: EnemyPool,
    relic_pool: RelicPool,
    hero: HeroData,
    current_floor: int = 1,
) -> None:
    """Apply all genome scalars to cloned game data (mutates in place)."""
    _apply_to_hero(genome, hero)
    _apply_to_cards(genome, card_pool, current_floor)
    _apply_to_enemies(genome, enemy_pool, current_floor)
    _apply_to_economy(genome, card_pool, relic_pool)


def apply_for_floor(
    genome: HierarchicalGenome,
    card_pool: CardPool,
    enemy_pool: EnemyPool,
    floor: int,
) -> None:
    """Apply genome for a specific floor (dynamic scaling)."""
    _apply_to_cards(genome, card_pool, floor)
    _apply_to_enemies(genome, enemy_pool, floor)


def _apply_to_hero(genome: HierarchicalGenome, hero: HeroData) -> None:
    hero.starting_health = max(30, round(hero.starting_health * genome.hero_health_scalar))
    hero.starting_gold = max(50, round(hero.starting_gold * genome.hero_start_gold_scalar))
    hero.starting_mana = max(2, min(4, hero.starting_mana + genome.hero_mana_offset))


def _apply_to_cards(genome: HierarchicalGenome, card_pool: CardPool, floor: int) -> None:
    for card in card_pool.cards_by_id.values():
        for action in card.actions:
            if action.type == ActionType.DealDamage:
                action.value = _calc_card_damage(genome, card, action.value, floor)
            elif action.type == ActionType.GainBlock:
                action.value = _calc_card_block(genome, card, action.value, floor)
            elif action.type == ActionType.ApplyStatusEffect:
                action.value = _calc_card_effect_stacks(genome, card, action.value)

        card.mana_cost = _calc_card_cost(genome, card, card.mana_cost)


def _calc_card_damage(genome, card, base_value: int, floor: int) -> int:
    scalar = genome.global_damage_multiplier
    scalar *= genome.card_type_scalars.get(card.type.value, 1.0)
    scalar *= genome.card_star_scalars.get(card.star_rating, 1.0)
    scalar *= genome.get_progression_scalar(floor, ScalingType.Damage)

    ovr = genome.card_damage_overrides.get(card.id)
    if ovr is not None:
        scalar *= ovr

    return max(1, round(base_value * scalar))


def _calc_card_block(genome, card, base_value: int, floor: int) -> int:
    scalar = genome.global_block_multiplier
    scalar *= genome.card_type_scalars.get(card.type.value, 1.0)
    scalar *= genome.card_star_scalars.get(card.star_rating, 1.0)
    scalar *= genome.get_progression_scalar(floor, ScalingType.Block)
    return max(1, round(base_value * scalar))


def _calc_card_effect_stacks(genome, card, base_value: int) -> int:
    scalar = genome.card_star_scalars.get(card.star_rating, 1.0)
    return max(1, round(base_value * scalar))


def _calc_card_cost(genome, card, base_cost: int) -> int:
    scalar = genome.global_mana_cost_multiplier
    ovr = genome.card_mana_cost_overrides.get(card.id)
    if ovr is not None:
        scalar *= ovr
    return max(0, round(base_cost * scalar))


def _apply_to_enemies(genome: HierarchicalGenome, enemy_pool: EnemyPool, floor: int) -> None:
    floor_diff = genome.get_floor_difficulty_multiplier(floor)

    for enemy in enemy_pool.enemies_by_id.values():
        health_scalar = genome.global_health_multiplier
        health_scalar *= genome.enemy_star_scalars.get(enemy.star_rating, 1.0)
        health_scalar *= genome.get_progression_scalar(floor, ScalingType.Health)
        health_scalar *= floor_diff

        hp_ovr = genome.enemy_health_overrides.get(enemy.id)
        if hp_ovr is not None:
            health_scalar *= hp_ovr

        enemy.starting_health = max(1, round(enemy.starting_health * health_scalar))

        for weighted_action in enemy.action_set:
            action = weighted_action.item
            if action.type == ActionType.DealDamage:
                dmg_scalar = genome.global_damage_multiplier
                dmg_scalar *= genome.enemy_star_scalars.get(enemy.star_rating, 1.0)
                dmg_scalar *= genome.get_progression_scalar(floor, ScalingType.Damage)
                dmg_scalar *= floor_diff
                action.value = max(1, round(action.value * dmg_scalar))
            elif action.type == ActionType.GainBlock:
                blk_scalar = genome.global_block_multiplier
                blk_scalar *= genome.enemy_star_scalars.get(enemy.star_rating, 1.0)
                blk_scalar *= genome.get_progression_scalar(floor, ScalingType.Block)
                action.value = max(1, round(action.value * blk_scalar))


def _apply_to_economy(
    genome: HierarchicalGenome,
    card_pool: CardPool,
    relic_pool: RelicPool,
) -> None:
    for star in range(1, 6):
        if star in card_pool.base_shop_costs:
            card_pool.base_shop_costs[star] = round(
                card_pool.base_shop_costs[star] * genome.global_gold_multiplier
            )
        if star in relic_pool.base_shop_costs:
            relic_pool.base_shop_costs[star] = round(
                relic_pool.base_shop_costs[star] * genome.global_gold_multiplier
            )
